sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -80,486 +65,209 @@ protected sealed override void Initialize(SonarAnalysisContext context)
+                                removableInternalMembers.Add(internalSymbol);
                             }
 
-                            var usedSymbols = new HashSet<ISymbol>();
-                            var emptyConstructors = new HashSet<ISymbol>();
+                            var usageCollector = new SymbolUsageCollector(
+                                c.Compilation.GetSemanticModel,
+                                removableSymbolsCollector.PrivateSymbols.Select(s => s.Name).ToHashSet());
 
-                            var propertyAccessorAccess = new Dictionary<IPropertySymbol, AccessorAccess>();
+                            VisitDeclaringReferences(namedType, usageCollector);
 
-                            CollectUsedSymbols(declarationCollector, usedSymbols, declaredPrivateSymbols, propertyAccessorAccess);
-                            CollectUsedSymbolsFromCtorInitializerAndCollectEmptyCtors(declarationCollector,
-                                usedSymbols, emptyConstructors);
-
-                            ReportIssues(cc, usedSymbols, declaredPrivateSymbols, emptyConstructors, fieldLikeSymbols);
-                            ReportUnusedPropertyAccessors(cc, usedSymbols, declaredPrivateSymbols, propertyAccessorAccess);
+                            var diagnostics = GetDiagnostics(usageCollector, removableSymbolsCollector.PrivateSymbols, "private",
+                                removableSymbolsCollector.FieldLikeSymbols);
+                            foreach (var diagnostic in diagnostics)
+                            {
+                                cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
+                            }
                         },
                         SymbolKind.NamedType);
 
                     c.RegisterCompilationEndAction(
-                       cc =>
-                       {
-                           if (!shouldRaise)
-                           {
-                               return;
-                           }
-
-                           foreach (var diagnostic in possibleUnusedInternalMembers)
-                           {
-                               cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
-                           }
-                       });
-                });
-        }
+                        cc =>
+                        {
+                            var foundInternalsVisibleTo = cc.Compilation.Assembly
+                                .GetAttributes(KnownType.System_Runtime_CompilerServices_InternalsVisibleToAttribute)
+                                .Any();
 
-        private static bool IsInternalVisibleToAttribute(AttributeSyntax attribute, SemanticModel semanticModel)
-        {
+                            if (foundInternalsVisibleTo)
+                            {
+                                return;
+                            }
+
+                            var usageCollector = new SymbolUsageCollector(

```

Comment:
Optimisation - no point in re-processing all named types if there are no removable internal types.